tools/nimdigger.nim

```diff
@@ -0,0 +1,327 @@
+template randMatrix*[T](m, n: int): Matrix[T] = randMatrix[T](m, n, T(1.0))
+let B = randMatrix[float32](20, 10)
+```
+```bash
+$ nim r tools/nimdigger.nim --oldnew:v0.19.0..v0.20.0 -- \
+  bin/nim c --hints:off --skipparentcfg --skipusercfg /tmp/t16376.nim
+fd16875561634e3ef24072631cf85eeead6213f2 is the first REGRESSION commit
+```
+
+## notes
+* this uses `git` (in particular `bisect`), `csources`, `csources_v`, `bash`, `make`/`gmake`
+* Unstable API, subject to change
+]##
+
+#[
+## TODO
+allow a way to verify that oldnew revisions honor what's implied by bisectBugfix:true|false
+
+## note
+we should give exit code = 125 to commits where nim won't build, to skip over, see also:
+https://stackoverflow.com/a/22592593/1426932 (Magic exit statuses)
+> anything above 127 makes the bisection fail with something like:
+> 125 is magic and makes the run be skipped with git bisect skip.
+]#
+
+import std/[os, osproc, strformat, macros, strutils, tables, algorithm]
+
+proc `$`(a: ref): string =
+  if a == nil: "nil" else: $a[]
+
+template dbg(args: varargs[untyped]): untyped =
+  # so users can swap in their own better logging until stdlib has one
+  echo args
+
+type
+  DiggerOpt = object ## nimdigger input
+    rev: string
+    nimDir: string
+    compileNim: bool
+    fetch: bool
+    csourcesBuildArgs: string
+    buildAllCsources: bool
+    verbose: bool
+
+    # bisect cmds
+    # TODO: allow user to not compile nim, for cases where it's not needed
+    oldnew: string # eg: v0.20.0~10..v0.20.0
+    bisectCmd: string # eg: bin/nim c --hints:off --skipparentcfg --skipusercfg $timn_D/tests/nim/all/t12329.nim 'arg1 bar' 'arg2'
+    bisectBugfix: bool
+  CsourcesOpt = ref object
+    url: string
+    dir: string
+    rev: string
+    binDir: string
+    csourcesBuildArgs: string
+    revs: seq[string]
+    fetch: bool
+    name: string
+    nimCsourcesExe: string
+  DiggerState = ref object ## nimdigger internal state
+    coptv0, coptv1: CsourcesOpt
+    binDir: string
+    nimDir: string
+    rev: string
+
+const
+  csourcesRevs = "v0.9.4 v0.13.0 v0.15.2 v0.16.0 v0.17.0 v0.17.2 v0.18.0 v0.19.0 v0.20.0 64e34778fa7e114b4afc753c7845dee250584167".split
+  csourcesV1Revs = "a8a5241f9475099c823cfe1a5e0ca4022ac201ff".split
+  NimDiggerEnv = "NIMDIGGER_HOME"
+  ExeExt2 = when ExeExt.len > 0: "." & ExeExt else: ""
+
+var verbose = false
+
+proc isSimulate(): bool =
+  defined(nimDiggerSimulate)
+
+proc runCmd(cmd: string) =
+  # TODO: allow `dir` param (or use `runCmdOutput`)
+  if isSimulate():
+    dbg cmd
+  else:
+    if verbose: dbg cmd
+    doAssert execShellCmd(cmd) == 0, cmd
+
+proc runCmdOutput(cmd: string, dir = ""): string =
+  if verbose: dbg cmd, dir
+  let (outp, status) = execCmdEx(cmd, workingDir = dir)
+  doAssert status == 0, indent(&"status: {status}\ncmd: {cmd}\ndir: {dir}\noutput: {outp}", 2)
+  result = outp
+  stripLineEnd(result)
+
+macro ctor(obj: untyped, a: varargs[untyped]): untyped =
```

Comment:
I feel that this is unnecessary - the amount of typing it saves is minimal compared to the decreased readability and increased mental parsing required. Also, it's only used once in the entire file.

Reply:
I'm also using this in https://github.com/nim-lang/fusion/pull/32 and intend to add it to std/sugar in subsequent PR which will refactor its different uses.

Reply:
The name should be `sugar.construct` IMO.

Reply:
I still feel that the tradeoff between writeability and readability limits its usefulness though.

Reply:
I don't understand this remark but we can discuss this in an RFC about how we should have `sugar.construct`.

Reply:
renamed to construct, RFC left TODO